src/plugins/s390.c

```diff
@@ -398,3 +399,353 @@ gchar* bd_s390_zfcp_sanitize_lun_input (const gchar *lun, GError **error) {
 
     return fulllun;
 }
+
+/**
+ * bd_s390_zfcp_online:
+ * @devno: zfcp device number
+ * @wwpn: zfcp WWPN (World Wide Port Number)
+ * @lun: zfcp LUN (Logical Unit Number)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether a zfcp device was successfully switched online
+ */
+gboolean bd_s390_zfcp_online (const gchar *devno, const gchar *wwpn, const gchar *lun, GError **error) {
+    gboolean boolrc = FALSE;
+    gint rc = 0;
+    FILE *fd = NULL;
+    DIR *pdfd = NULL;
+    const gchar *zfcp_cio_free[4] = {"/usr/sbin/zfcp_cio_free", "-d", devno, NULL};
+    const gchar *chccwdev[4] = {"/usr/sbin/chccwdev", "-e", devno, NULL};
+
+    gchar *zfcpsysfs = "/sys/bus/ccw/drivers/zfcp";
+    gchar *online = g_strdup_printf ("%s/%s/online", zfcpsysfs, devno);
+    gchar *portdir = g_strdup_printf ("%s/%s/%s", zfcpsysfs, devno, wwpn);
+    gchar *unitadd = g_strdup_printf ("%s/unit_add", portdir);
+    gchar *unitdir = g_strdup_printf ("%s/%s", portdir, lun);
+    gchar *failed = g_strdup_printf ("%s/failed", unitdir);
+
+    /* part 01: make sure device is available/not on device ignore list */
+    fd = fopen (online, "r");
+    if (!fd) {
+        boolrc = bd_utils_exec_and_report_error (zfcp_cio_free, NULL, error);
+
+        if (!boolrc) {
+            fclose (fd);
+            g_free (online);
+            g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                         "Could not remove device %s from device ignore list.", devno);
+            return FALSE;
+        }
+        /* fd is NULL at this point, so try to open it again */
+        fd = fopen (online, "r");
+        g_free (online);
+
+        /* still no luck, fail */
+        if (!fd) {
+            g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                         "Could not open device %s even after removing it from" " the device ignore list.", devno);
+            return FALSE;
+        }
+    }
+
+    /* part 02: check to make sure/turn device online */
+    rc = fgetc (fd);
+    if (rc == EOF) {
+        /* there was some error checking the 'online' status */
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_IO,
+                     "Error checking if device %s is online", devno);
+        fclose (fd);
+        return FALSE;
+    }
+    if (rc == 1) {
+        /*
+         * otherwise device's status indicates that it's arleady online, so
+         * just close the fd and proceed; we don't return because although 'online'
+         * status may be correct, the device may not be completely online and ready
+         * to use just yet
+         */
+        fclose (fd);
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                     "Device %s is already online", devno);
```

Comment:
So this function may return ``TRUE`` and an error at the same time? That's not really nice and it would result in an exception being raised in Python for a successful run of the function.